### PR TITLE
Add the LLVM IR unreachable instruction to the LLVMIR dialect.

### PR DESCRIPTION
**Pull request by schweitzpgi**
http://llvm.org/docs/LangRef.html#unreachable-instruction

This instruction will be to support (otherwise regular) calls to the Flang
runtime that are known to never return to the caller.

### CODE REVIEW

**Comment by schweitzpgi** [comment]
Nvidia has signed the CLA.

@googlebot I signed it!

File: include/mlir/LLVMIR/LLVMOps.td [hunk 1]
```diff
@@ -300,6 +300,16 @@ def LLVM_ReturnOp : LLVM_TerminatorOp<"return", []> {
   let parser = [{ return parseReturnOp(parser, result); }];
   let printer = [{ printReturnOp(p, *this); }];
 }
+def LLVM_UnreachableOp : LLVM_TerminatorOp<"unreachable", []> {
+  string llvmBuilder = [{
+    builder.CreateUnreachable();
+  }];
+  let verifier = [{
+    return success();
```
Comment (River707) [inline include/mlir/LLVMIR/LLVMOps.td]: I don't think it's necessary to define the verifier if you don't need to.

File: lib/LLVMIR/IR/LLVMDialect.cpp [hunk 2]
```diff
@@ -635,6 +635,10 @@ static void printReturnOp(OpAsmPrinter *p, ReturnOp &op) {
   *p << ' ' << *op.getOperand(0) << " : " << op.getOperand(0)->getType();
 }
 
+static void printUnreachableOp(OpAsmPrinter *p, UnreachableOp &op) {
+  *p << op.getOperationName() << " : () -> ()";
```
Comment (River707) [inline lib/LLVMIR/IR/LLVMDialect.cpp]: Don't print a trailing empty functional type, it isn't necessary.

Comment (River707) [reply lib/LLVMIR/IR/LLVMDialect.cpp]: On that note, can we just inline this function into the op definition? It is small enough.

File: lib/LLVMIR/IR/LLVMDialect.cpp [hunk 3]
```diff
@@ -653,6 +657,15 @@ static ParseResult parseReturnOp(OpAsmParser *parser, OperationState *result) {
   return success();
 }
 
+// <operation> ::= `llvm.unreachable` `:` type-list-no-parens
+static ParseResult parseUnreachableOp(OpAsmParser *parser,
+                                      OperationState *result) {
+  Type type;
```
Comment (River707) [inline lib/LLVMIR/IR/LLVMDialect.cpp]: This parser shouldn't need to do anything, just inline it into the op definition.

**Comment by schweitzpgi** [comment]
Thanks for the review, River. I've made your requested changes.

**Comment by River707** [comment]
Can you add a test to "test/Target/llvmir.mlir" to test the conversion to LLVM IR?

**Comment by schweitzpgi** [comment]
> Can you add a test to "test/Target/llvmir.mlir" to test the conversion to LLVM IR?

Sure.

File: test/Target/llvmir.mlir [hunk 4]
```diff
@@ -825,3 +825,9 @@ func @stringconstant() -> !llvm<"i8*"> {
   llvm.return %1 : !llvm<"i8*">
 }
 
+// CHECK-LABEL: define void @noreach() {
+// CHECK-NEXT:    unreachable
```
Comment (River707) [inline test/Target/llvmir.mlir]: We don't need to check the trailing '}' here.

**Comment by joker-eph** [comment]
CLA overridden: contribution from Nvidia.

File: test/Target/llvmir.mlir (hunk 4 above)
Comment (schweitzpgi) [reply test/Target/llvmir.mlir]: I see your point, however I left it in there to be consistent with some similar checks earlier in the file.

Comment (River707) [reply test/Target/llvmir.mlir]: It's alright to diverge here, we can clean the rest in a followup.

Comment (schweitzpgi) [reply test/Target/llvmir.mlir]: It looks like these last few tests cherry pick the LLVM IR output. If that's the new preferred pattern, perhaps I should remove the CHECK-LABEL line as well.

**Pull request closed without merging**